src/pdb/string.rs

```diff
@@ -0,0 +1,271 @@
+                    // we might need to do some additional filtering here
+                    content: LongBody::Ucs2le(string.encode_utf16().collect()),
+                }))
+            }
+        } else {
+            Err(StringError::TooLong)
+        }
+    }
+    /// Creates a DeviceSQLString containing an ISRC instead of an expected string
+    ///
+    /// The DeviceSQL Database as used by rekordbox has a strange oddity
+    /// where it serializes strings containing a tracks ISRC (International
+    /// Standard Recording Code) in an unexpected format, if this is desired,
+    /// use this constructor function instead of [DeviceSQLString::new]
+    pub fn new_isrc(string: String) -> Result<Self, StringError> {
+        if string.is_empty() {
+            return Ok(Self::empty());
+        }
+        // Basic validation taken from:
+        // https://isrc.ifpi.org/downloads/ISRC_Bulletin-2015-01.pdf
+        if string.len() != 12 || !string.is_ascii() {
+            return Err(StringError::InvalidISRC);
+        }
+        Ok(Self(DeviceSQLStringImpl::Long {
+            content: LongBody::Isrc(NullString::from_string(string)),
+        }))
+    }
+    /// Extract the Rust string from the DeviceSQLString.
+    ///
+    /// Consumes itself in the process.
+    pub fn into_string(self) -> Result<String, StringError> {
+        match self.0 {
+            DeviceSQLStringImpl::ShortASCII { content: vec, .. }
+            | DeviceSQLStringImpl::Long {
+                content: LongBody::Ascii(vec),
+                ..
+            } => String::from_utf8(vec).map_err(|_| StringError::Encoding),
+            DeviceSQLStringImpl::Long {
+                content: LongBody::Isrc(str),
+                ..
+            } => str
+                .into_string_lossless()
+                .map_err(|_| StringError::Encoding),
+            DeviceSQLStringImpl::Long {
+                content: LongBody::Ucs2le(vec),
+            } => String::from_utf16(&vec).map_err(|_| StringError::Encoding),
+        }
+    }
+    /// Create an empty DeviceSQLString
+    ///
+    /// Should be used to construct known empty strings.
+    #[must_use]
+    pub const fn empty() -> Self {
+        Self(DeviceSQLStringImpl::ShortASCII {
+            content: Vec::new(),
+        })
+    }
+}
+
+/// A String encapsulating how DeviceSQL VARCHAR's are structured
+///
+/// This implementation forces them to be immutable for now
+#[binrw]
+#[derive(Debug, PartialEq)]
+enum DeviceSQLStringImpl {
+    /// Short-String optimization case
+    ShortASCII {
+        #[br(temp, assert(header & 0b1 == 1))]
+        #[bw(calc = (((content.len() + 1) << 1) | 1) as u8)]
+        header: u8,
+
+        #[br(count = (header >> 1) - 1)]
+        content: Vec<u8>,
+    },
+    /// Regular long form strings, containing possibly different encodings
+    Long {
+        #[br(temp)]
+        #[br(assert(flags & 0b1 == 0))]
```

Comment:
Why `& 0b1`? Shouldn't it be `flags == 0x40 || flags == 0x90`?

Reply:
Well, yes in the reference documentation but I recently gained some insight into DeviceSQLStrings that changed my view of them slightly. See the corresponding PR: https://github.com/Deep-Symmetry/crate-digger/pull/26/files

Reply:
To summarize, the first byte serves doubly duty, it is either a 32-bit bitmask or an 8-bit bitmask. To differentiate between the two, the LSB of the first byte can be tested. This is what the asserts are doing here. In the 8-bit case, the upper 7 seven bits contain the length of the short ascii string and in the 32-bit case, the upper seven bits contain information about the contained string (probably encoding information). See [Greg Kemnitz blog post](https://bufferpool.quora.com/What-does-a-db-engine-string-look-like) for some more insights (greg is the author of the Encirq version of DeviceSQL). In the linked PR, I described a mapping of these bits to string metadata, but we can't be sure they are accurate them until we see other types of strings (which I think is unlikely).

Reply:
So the only requirement for being a valid `Long` string, is `flags & 0b1 == 0`, the more indepth check is done in the `LongBody` enum were I `preassert` the flags to actually find the correct enum variant. This solution makes it easy to add new string variants if we ever encounter them.

Reply:
Could you add a comment with your explanations? That will make it easier to understand why we check the LSB.

Reply:
Sure